File: Palace/Views/TPPLinearView.m
#import "TPPLinearView.h"

@interface TPPLinearView ()

@property (nonatomic) CGFloat minimumRequiredHeight;
@property (nonatomic) CGFloat minimumRequiredWidth;

@end

@implementation TPPLinearView

#pragma mark UIView

- (void)layoutSubviews
{
  CGFloat x = 0.0;
  
  for(UIView *const view in self.subviews) {
    CGFloat const w = CGRectGetWidth(view.frame);
    CGFloat const h = CGRectGetHeight(view.frame);
    CGFloat y;
    switch(self.contentVerticalAlignment) {
      case TPPLinearViewContentVerticalAlignmentTop:
        y = 0;
        break;
      case TPPLinearViewContentVerticalAlignmentMiddle:
        y = round((CGRectGetHeight(self.frame) - h) / 2.0);
        break;
      case TPPLinearViewContentVerticalAlignmentBottom:
        y = CGRectGetHeight(self.frame) - h;
        break;
    }
    view.frame = CGRectMake(x, y, w, h);
    self.minimumRequiredWidth = x + w;
    self.minimumRequiredHeight = h > self.minimumRequiredHeight ? h : self.minimumRequiredHeight;
    x += w + self.padding;
  }
}

- (CGSize)sizeThatFits:(CGSize)size
{
  [self layoutIfNeeded];
  
  CGFloat const w = self.minimumRequiredWidth;
  CGFloat const h = self.minimumRequiredHeight;
  
  if(CGSizeEqualToSize(size, CGSizeZero)) {
    return CGSizeMake(w, h);
  }
  
  return CGSizeMake(w > size.width ? size.width : w, h > size.height ? size.height : h);
}

#pragma mark -

- (void)setContentVerticalAlignment:
(TPPLinearViewContentVerticalAlignment const)contentVerticalAlignment
{
  _contentVerticalAlignment = contentVerticalAlignment;
  
  [self setNeedsLayout];
}

- (void)setPadding:(CGFloat const)padding
{
  _padding = padding;
  
  [self setNeedsLayout];
}

@end

File: Palace/Views/TPPLinearView.swift
import UIKit

@objc enum TPPLinearViewContentVerticalAlignment: Int {
  case top
  case middle
  case bottom
}

@objc class TPPLinearView: UIView {

  @objc var contentVerticalAlignment: TPPLinearViewContentVerticalAlignment = .top {
    didSet { setNeedsLayout() }
  }

  @objc var padding: CGFloat = 0 {
    didSet { setNeedsLayout() }
  }

  private var minimumRequiredHeight: CGFloat = 0
  private var minimumRequiredWidth: CGFloat = 0

  override func layoutSubviews() {
    var x: CGFloat = 0

    for view in subviews {
      let w = view.frame.width
      let h = view.frame.height
      let y: CGFloat

      switch contentVerticalAlignment {
      case .top:
        y = 0
      case .middle:
        y = round((frame.height - h) / 2.0)
      case .bottom:
        y = frame.height - h
      }

      view.frame = CGRect(x: x, y: y, width: w, height: h)
      minimumRequiredWidth = x + w
      minimumRequiredHeight = max(h, minimumRequiredHeight)
      x += w + padding
    }
  }

  override func sizeThatFits(_ size: CGSize) -> CGSize {
    layoutIfNeeded()

    let w = minimumRequiredWidth
    let h = minimumRequiredHeight

    if size == .zero {
      return CGSize(width: w, height: h)
    }

    return CGSize(
      width: min(w, size.width),
      height: min(h, size.height)
    )
  }
}
